src/puffin/src/puffin_manager/cache_manager/moka_cache_manager.rs

```diff
@@ -0,0 +1,941 @@
+                        guard: Arc::new(FsBlobGuard {
+                            path,
+                            delete_queue: self.delete_queue.clone(),
+                        }),
+                        size: meta.len(),
+                    };
+                    // A duplicate file will be moved to the delete queue.
+                    let _dup_file = elems.insert(key, v);
+                }
+            }
+        }
+
+        for (key, value) in elems {
+            self.cache.insert(key, value).await;
+        }
+
+        Ok(())
+    }
+
+    /// Walks through the directory and calculate the total size of all files in the directory.
+    async fn get_dir_size(path: &PathBuf) -> Result<u64> {
+        let mut size = 0;
+        let mut wd = WalkDir::new(path).filter(|entry| async move {
+            match entry.file_type().await {
+                Ok(ft) if ft.is_dir() => Filtering::Ignore,
+                _ => Filtering::Continue,
+            }
+        });
+
+        while let Some(entry) = wd.next().await {
+            let entry = entry.context(WalkDirSnafu)?;
+            size += entry.metadata().await.context(MetadataSnafu)?.len();
+        }
+
+        Ok(size)
+    }
+
+    async fn delete_routine(mut receiver: Receiver<DeleteTask>) {
+        while let Some(task) = receiver.recv().await {
+            match task {
+                DeleteTask::File(path) => {
+                    if let Err(err) = fs::remove_file(&path).await {
+                        if err.kind() == std::io::ErrorKind::NotFound {
+                            continue;
+                        }
+
+                        warn!(err; "Failed to remove the file.");
+                    }
+                }
+                DeleteTask::Dir(path) => {
+                    let deleted_path = path.with_extension(DELETED_EXTENSION);
+                    if let Err(err) = fs::rename(&path, &deleted_path).await {
+                        if err.kind() == std::io::ErrorKind::NotFound {
+                            continue;
+                        }
+
+                        // Remove the deleted directory if the rename fails and retry
+                        let _ = fs::remove_dir_all(&deleted_path).await;
+                        if let Err(err) = fs::rename(&path, &deleted_path).await {
+                            warn!(err; "Failed to rename the dangling directory to deleted path.");
+                            continue;
+                        }
+                    }
+                    if let Err(err) = fs::remove_dir_all(&deleted_path).await {
+                        warn!(err; "Failed to remove the dangling directory.");
+                    }
+                }
+                DeleteTask::Terminate => {
+                    break;
+                }
+            }
+        }
+    }
+}
+
+impl Drop for MokaCacheManager {
+    fn drop(&mut self) {
+        let _ = self.delete_queue.try_send(DeleteTask::Terminate);
```

Comment:
Maybe we don't need this task as the receiver will return `None` once the sender is dropped.

Reply:
Here is the initiative to let `delete_routine` exit, otherwise the `*Guard` in the cache will delete the file when the `MokaCacheManager` drops

Reply:
Got it. Could you please document this in the comment?